src/iarc7_motion/iarc_tasks/track_roomba_task.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+
+import math
+import rospy
+import tf2_ros
+import tf2_geometry_msgs
+
+from geometry_msgs.msg import TwistStamped
+from geometry_msgs.msg import Point
+from geometry_msgs.msg import PointStamped
+
+
+from iarc7_msgs.msg import OdometryArray
+
+from .abstract_task import AbstractTask
+from iarc_tasks.task_states import (TaskRunning,
+                                    TaskDone,
+                                    TaskCanceled,
+                                    TaskAborted,
+                                    TaskFailed)
+from iarc_tasks.task_commands import (VelocityCommand,
+                                      ArmCommand,
+                                      NopCommand)
+
+from height_holder import HeightHolder
+
+class TrackObjectTaskState:
+    init = 0
+    track = 1
+    waiting = 2
+
+class TrackRoombaTask(AbstractTask):
+
+    def __init__(self, task_dictionary):
+
+        self._roomba_id = task_dictionary['frame_id']
+
+        self._roomba_id = self._roomba_id  + '/base_link'
+        
+        #p-terms
+        self._k_x = .35
+        self._k_y = .35
+
+        if self._roomba_id is None or len(self._roomba_id) < 2:
+            raise ValueError('An invalid or null roomba id was provided')
+
+        self._roomba_odometry = None
+        self._roomba_array = None
+        self._roomba_point = None
+
+        self._roomba_found = False
+        self._canceled = False
+
+        self._tf_buffer = tf2_ros.Buffer()
+        self._tf_listener = tf2_ros.TransformListener(self._tf_buffer)  
+
+        self._roomba_status_sub = rospy.Subscriber(
+            'roombas', OdometryArray, 
+            self._receive_roomba_status)
+
+        try:
+            self._TRANSFORM_TIMEOUT = rospy.get_param('~transform_timeout')
+            self._MIN_MANEUVER_HEIGHT = rospy.get_param('~min_maneuver_height')
+            self._TRACK_HEIGHT = rospy.get_param('~track_roomba_height')
+            self._MAX_TRANSLATION_SPEED = rospy.get_param('~max_translation_speed')
+            self._MAX_ACCEPTABLE_DIST = rospy.get_param('~roomba_max_acceptable_dist')
+        except KeyError as e:
+            rospy.logerr('Could not lookup a parameter for track roomba task')
+            raise
+
+        if self._TRACK_HEIGHT < self._MIN_MANEUVER_HEIGHT:
+            raise ValueError('Track Roomba height was below the minimum maneuver height')
+
+        self._z_holder = HeightHolder()
+
+        self._state = TrackObjectTaskState.init
+
+        while self._roomba_array is None:
+            pass
+
+    def _receive_roomba_status(self, data):
+        self._roomba_array = data
+
+    def get_desired_command(self):
+        if (self._state == TrackObjectTaskState.init or 
+            self._state == TrackObjectTaskState.waiting):
```

Comment:
This whole section had wacky logic because I was always changing it so I refactored it.